modules/core/src/smithy4s/Service.scala

```diff
@@ -187,4 +187,66 @@ object Service {
     final def toPolyFunction[P[_, _, _, _, _]](algebra: PolyFunction5[Op, P]): PolyFunction5[Op, P] = algebra
     final def mapK5[F[_, _, _, _, _], G[_, _, _, _, _]](algebra: PolyFunction5[Op, F], function: PolyFunction5[F, G]): PolyFunction5[Op, G] = algebra.andThen(function)
   }
+
+  object Builder {
+    def fromService[Alg[_[_, _, _, _, _]]](
+        service: Service[Alg]
+    ): Builder[Alg, service.Operation] =
+      new Builder[Alg, service.Operation](service, PolyFunction5.identity)
+  }
+
+  final case class Builder[Alg[_[_, _, _, _, _]], Op[_, _, _, _, _]] private(
+      private val base: Service.Aux[Alg, Op],
+      private val endpointMapper: PolyFunction5[Endpoint.ForOperation[Op]#e, Endpoint.ForOperation[Op]#e]
+  ) {
+
+    def mapEndpointEach(
+        mapper: PolyFunction5[Endpoint.ForOperation[Op]#e, Endpoint.ForOperation[Op]#e]
+    ): Builder[Alg, Op] = copy(
+      endpointMapper = (endpointMapper.andThen(mapper)).unsafeCacheBy(
+        base.endpoints.map(Kind5.existential(_)),
+        identity
+      )
+    )
+
+    def build: Service.Aux[Alg, Op] = new Service[Alg] {
+
+      override type Operation[I, E, O, SI, SO] = Op[I, E, O, SI, SO]
+
+      override val endpoints: List[Endpoint[_, _, _, _, _]] =
+        // Explicit upcast to help Scala 3's type inference
+        base.endpoints.map(e => endpointMapper(e): Endpoint[_, _, _, _, _])
+
+      override def endpoint[I, E, O, SI, SO](
+          op: Op[I, E, O, SI, SO]
+      ): (I, Endpoint[I, E, O, SI, SO]) = {
+        val (input, baseEndpoint) = base.endpoint(op)
+
+        (input, endpointMapper(baseEndpoint))
+      }
+
+      // todo: these should be easier ones
+      override val id: ShapeId = ???
+
+      override val version: String = ???
+
+      override val hints: Hints = ???
```

Comment:
@GaryAghedo can you add methods for these? It'd be the usual, just like everything in the endpoint builder.

Reply:
nice!
Yes sure.. I will add the code now